File: modules/common-net/src/test/java/com/shterneregen/securelan/common/net/transport/LineTextChannelTest.java
package com.shterneregen.securelan.common.net.transport;

import org.junit.jupiter.api.Test;

import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.FutureTask;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class LineTextChannelTest {
    @Test
    void shouldExchangeUtf8Lines() throws Exception {
        try (ServerSocket serverSocket = new ServerSocket(0)) {
            FutureTask<String> serverTask = new FutureTask<>(() -> {
                try (Socket socket = serverSocket.accept(); LineTextChannel channel = new LineTextChannel(socket)) {
                    String received = channel.readLine();
                    channel.writeLine("reply: " + received);
                    return channel.remoteAddress();
                }
            });
            Thread serverThread = new Thread(serverTask, "line-text-channel-test-server");
            serverThread.start();

            try (Socket socket = new Socket("127.0.0.1", serverSocket.getLocalPort());
                 LineTextChannel channel = new LineTextChannel(socket)) {
                channel.writeLine("Привет SecureLanSuite");
                assertEquals("reply: Привет SecureLanSuite", channel.readLine());
            }

            String remoteAddress = serverTask.get();
            assertTrue(remoteAddress.contains("127.0.0.1") || remoteAddress.contains("localhost"));
        }
    }
}

File: modules/common-net/src/test/kotlin/com/shterneregen/securelan/common/net/transport/LineTextChannelTest.kt
package com.shterneregen.securelan.common.net.transport

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.net.ServerSocket
import java.net.Socket
import java.util.concurrent.FutureTask

class LineTextChannelTest {
    @Test
    fun shouldExchangeUtf8Lines() {
        ServerSocket(0).use { serverSocket ->
            val serverTask = FutureTask<String> {
                serverSocket.accept().use { socket ->
                    LineTextChannel(socket).use { channel ->
                        val received = channel.readLine()
                        channel.writeLine("reply: $received")
                        channel.remoteAddress()
                    }
                }
            }
            Thread(serverTask, "line-text-channel-test-server").start()

            Socket("127.0.0.1", serverSocket.localPort).use { socket ->
                LineTextChannel(socket).use { channel ->
                    channel.writeLine("Привет SecureLanSuite")
                    assertEquals("reply: Привет SecureLanSuite", channel.readLine())
                }
            }

            val remoteAddress = serverTask.get()
            assertTrue(remoteAddress.contains("127.0.0.1") || remoteAddress.contains("localhost"))
        }
    }
}
